### PR TITLE
[low code connectors] generate complete json schema from classes

### PR DESCRIPTION
**Pull request by brianjlai**
## What
Adds a new method for generating the complete json schema of all the components of the low code declarative language starting from the top level `CheckStream` and `DeclarativeStream`s. How this is ultimately surfaced to the user is an open question, but for now it's returned as a flat string which can be formatted on other apps like Sublime.

## How
Performs a complete traversal of the object model starting at the top of the declarative framework by traversing through each component relationship and expanding fields that are represented by an interface into their concrete class implementations. 

This uses a recursive algorithm to perform the traversal by iterating over a class' underlying fields. For each field we unpack them if they're stored in a generic data structure and translate them into the class types that implement the interface. Then we perform the same process on the underlying declarative components (all other types and primitives are ignored)

After the traversal and translating the interfaces into classes, the `dataclasses-jsonschema` is used to generate a schema from the data model which is then transformed into a json-readable format

## Recommended reading order
1. `yaml_declarative_source.py`

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py [hunk 1]
```diff
@@ -79,3 +82,74 @@ def _stream_configs(self):
...
+        # Recursive base case to stop recursion if we have already expanded an interface in case of cyclical components
+        # like CompositeErrorHandler
+        if expand_class.__name__ in visited:
+            return visited[expand_class.__name__]
+        visited[expand_class.__name__] = expand_class
+
+        next_classes = []
+        copy_cls = type(expand_class.__name__, expand_class.__bases__, dict(expand_class.__dict__))
+        class_fields = fields(copy_cls)
+        for field in class_fields:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:112]: This part of the code follows a similar structure to the original schema validation in #15543 where we unpack the underlying types and replace the annotations

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py [hunk 2]
```diff
@@ -79,3 +82,74 @@ def _stream_configs(self):
...
+            unpacked_field_types = DeclarativeComponentFactory.unpack(field.type)
+            copy_cls.__annotations__[field.name] = unpacked_field_types
+            next_classes.extend(YamlDeclarativeSource._get_next_expand_classes(field.type))
+
+        for next_class in next_classes:
+            YamlDeclarativeSource.expand_schema_interfaces(next_class, visited)
+        return copy_cls
+
+    @staticmethod
+    def _get_next_expand_classes(field_type) -> list[type]:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:122]: We can just return a flat list of next components that need to be traversed because we are relying on unpack to properly resolve interfaces and retain the existing structure. This function is solely responsible for figuring out the next set of components that need to be expanded as part of the next recursive call.

File: airbyte-cdk/python/unit_tests/sources/declarative/test_factory.py [hunk 3]
```diff
@@ -807,3 +808,8 @@ def test_validate_types_nested_in_list():
 def test_unpack(test_name, input_type, expected_unpacked_types):
     actual_unpacked_types = DeclarativeComponentFactory.unpack(input_type)
     assert actual_unpacked_types == expected_unpacked_types
+
+
+def test_complete_schema():
```
Comment (girarda) [inline airbyte-cdk/python/unit_tests/sources/declarative/test_factory.py:813]: do we want to keep this test?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py [hunk 4]
```diff
@@ -79,3 +82,74 @@ def _stream_configs(self):
...
+                next_classes.extend(YamlDeclarativeSource._get_next_expand_classes(underlying_type))
+            return next_classes
+        return []
+
+
+class SchemaEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, property):
+            return str(obj)
+        elif isinstance(obj, HttpMethod):
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:153]: is this because `HttpMethod` is an enum? Is `RequestOptionType` somehow not a problem?

Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:156]: Yep this avoids the serialization issue for the `HttpMethod` enum. The reason why we need it for `HttpMethod` is because we assign the enum value `HttpMethod.GET` as the default so the enum (not just a string) shows up in the schema [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py#L44).

But for `RequestOptionType` since we don't assign the enum anywhere, enum class shows up in the schema, but not the enum value. That's why we didn't run into the serialization problem

File: airbyte-cdk/python/unit_tests/sources/declarative/test_factory.py (hunk 3 above)
Comment (brianjlai) [reply airbyte-cdk/python/unit_tests/sources/declarative/test_factory.py:813]: nope, this is just for my convenience testing, i'll remove it if we're just gonna generate the json file and put it into the repo

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py (hunk 4 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:156]: can we check if the type is an enum instead so we don't have to add another case to this method when we add another enum with a default value?

Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:156]: good point, will change this

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py [hunk 5]
```diff
@@ -84,3 +87,72 @@ def _stream_configs(self):
...
+        it will return the unpacked classes. Any non-declarative types will be skipped.
+        :param field_type: A field type that
+        :return:
+        """
+        generic_type = typing.get_origin(field_type)
+        if generic_type is None:
+            module = field_type.__module__
+            # We can only continue parsing declarative components since we explicitly inherit from the JsonSchemaMixin class which is
+            # used to generate the final json schema
+            if "airbyte_cdk.sources.declarative" in module and not isinstance(field_type, EnumMeta):
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:138]: That should work too. It does have an interesting side effect. Maybe it's what we want, but when interfaces like `PaginationStrategy` inherit `JsonSchemaMixin`, they show up in the schema. It might be what we want since they are part of the language, but also can be a little confusing because they don't have any fields or schemas to validate. They also turn every type into unions of the interface and concrete types

Basically the readability of the schema is worse, but it is _technically_ more accurate.

[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:138]: do they show up as a union of the subtypes or as an empty object? It think showing them is better if they're defined as the union, but worse otherwise

[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:138]: It'll look something like this when you view the ListStreamSlicer in the `definitions`. It is kind of the union, where the first referes the empty StreamSlicer object, and the second is the concrete definition which will actually get used in the validation.

```
ListStreamSlicer: [
    {
        "$ref": "#/definitions/StreamSlicer"
    },
    {
        "type": "object",
        "required":
        [
            "slice_values",
            "cursor_field",
            "config"
        ],
        "properties":
        {
            "slice_values":
            {
                "anyOf":
                [
                    {
                        "type": "array",
                        "items":
                        {
                            "type": "string"
                        }
                    },
                    {
                        "type": "string"
                    }
                ]
            },
            "cursor_field":
            {
                "anyOf":
                [
                    {
                        "$ref": "#/definitions/InterpolatedString"
                    },
                    {
                        "type": "string"
                    }
                ]
            },
            "config":
            {
                "type": "object"
            },
            "request_option":
            {
                "$ref": "#/definitions/RequestOption"
            }
        }
    }
]
```

[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/yaml_declarative_source.py:138]: discussed this on slack a bit, a slightly more verbose schema is probably acceptable vs module path assertions

**Review by girarda — APPROVED** [review]
approved pending the enum type check on https://github.com/airbytehq/airbyte/pull/15647/files#diff-f065e5627510a5d82633b9ac35383b6065f9b09ab4381f667763ad322a13dccaR156
